azblob/zc_retry_reader.go

```diff
@@ -45,19 +64,37 @@ type RetryReaderOptions struct {
 // through reading from the new response.
 type retryReader struct {
 	ctx             context.Context
-	response        *http.Response
 	info            HTTPGetterInfo
 	countWasBounded bool
 	o               RetryReaderOptions
 	getter          HTTPGetter
+
+	// we support Close-ing during Reads (from other goroutines), so we protect the shared state, which is response
+	responseMu       *sync.Mutex
+	response         *http.Response
 }
 
 // NewRetryReader creates a retry reader.
 func NewRetryReader(ctx context.Context, initialResponse *http.Response,
 	info HTTPGetterInfo, o RetryReaderOptions, getter HTTPGetter) io.ReadCloser {
-	return &retryReader{ctx: ctx, getter: getter, info: info, countWasBounded: info.Count != CountToEnd, response: initialResponse, o: o}
+	return &retryReader{
+		ctx: ctx,
+		getter: getter,
+		info: info,
+		countWasBounded: info.Count != CountToEnd,
+		response:    initialResponse,
+		responseMu:  &sync.Mutex{},
+		o:           o}
 }
 
+
+func (s *retryReader) setResponse(r *http.Response){
+	s.responseMu.Lock()
+	defer s.responseMu.Unlock()
+	s.response = r
+}
```

Comment:
Mino: Please ensure format is good, e.g. additional blank lines are avoided.

Reply:
Done. (Ran gofmt, which I thought had run before, but clearly it hadn't).